src/main/java/net/minecraftforge/registries/ForgeRegistryTagManager.java

```diff
@@ -0,0 +1,133 @@
+
+    ForgeRegistryTagManager(ForgeRegistry<V> owner)
+    {
+        this.owner = owner;
+    }
+
+    void bind(Map<TagKey<V>, HolderSet.Named<V>> holderTags)
+    {
+        IdentityHashMap<TagKey<V>, ITag<V>> newTags = new IdentityHashMap<>(this.tags);
+        Set<TagKey<V>> bound = new HashSet<>();
+
+        holderTags.forEach((key, holderSet) -> {
+            bound.add(key);
+            ((ForgeRegistryTag<V>) newTags.computeIfAbsent(key, ForgeRegistryTag::new)).bind(holderSet);
+        });
+
+        // Forcefully unbind any tags that didn't get bound
+        Sets.difference(this.tags.keySet(), bound).forEach(key -> ((ForgeRegistryTag<V>) newTags.get(key)).bind(null));
+
+        this.tags = newTags;
+    }
+
+    @NotNull
+    @Override
+    public ITag<V> getTag(@NotNull TagKey<V> name)
+    {
+        Objects.requireNonNull(name);
+        ITag<V> tag = this.tags.get(name);
+
+        if (tag == null)
+        {
+            // Create empty tag
+            tag = new ForgeRegistryTag<>(name);
+
+            // Mojang uses volatile and sets the tag map this way to not have the performance penalties of synced read access.
+            // However, this can generate a lot of new maps. We should to look into performance alternatives.
+            IdentityHashMap<TagKey<V>, ITag<V>> map = new IdentityHashMap<>(this.tags);
+            map.put(name, tag);
+            this.tags = map;
+        }
+
+        return tag;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @NotNull
+    @Override
+    public Optional<IReverseTag<V>> getReverseTag(@NotNull V value)
+    {
+        Objects.requireNonNull(value);
+        // All Holders are implementors of IReverseTag
+        return (Optional<IReverseTag<V>>) (Optional) this.owner.getHolder(value);
+    }
+
+    @Override
+    public boolean isKnownTagName(@NotNull TagKey<V> name)
+    {
+        Objects.requireNonNull(name);
+        ITag<V> tag = this.tags.get(name);
+        return tag != null && tag.isBound();
+    }
+
+    @NotNull
+    @Override
+    public Iterator<ITag<V>> iterator()
+    {
+        return Iterators.unmodifiableIterator(this.tags.values().iterator());
+    }
+
+    @NotNull
+    @Override
+    public Stream<ITag<V>> stream()
+    {
+        return this.tags.values().stream();
+    }
+
+    @NotNull
+    @Override
+    public Stream<TagKey<V>> getTagNames()
+    {
+        return this.tags.keySet().stream();
+    }
+
+    @NotNull
+    @Override
+    public TagKey<V> createTagKey(@NotNull ResourceLocation location)
+    {
+        Objects.requireNonNull(location);
+        return TagKey.create(this.owner.getRegistryKey(), location);
+    }
+
+    @NotNull
+    @Override
+    public TagKey<V> createOptionalTagKey(@NotNull ResourceLocation location, @NotNull Set<Supplier<V>> defaults)
+    {
+        TagKey<V> tagKey = createTagKey(location);
+
+        this.owner.getHolderHelper().ifPresent(h -> h.addOptionalTag(tagKey, defaults));
+
+        return tagKey;
```

Comment:
Part of me wonders if this method should have some sort of "failure" return for if it was unable to add an optional tag (such as if there is no holder helper), rather than just returning a new tag key. (This might not make sense to do, just figured I would ask as it did cross my mind)

Reply:
Failure return or failure exception?

Reply:
Not sure of what the best would be. Maybe an optional or null? Might get a bit messy or even maybe an exception though that might not make sense. This is in part why I was mentioning it as I wonder if it makes sense to do something here, but I don't know *what*.